src/main/kotlin/uk/gov/justice/digital/hmpps/hmppsbookavideolinkapi/service/locations/availability/LocationAvailabilityService.kt

```diff
@@ -0,0 +1,94 @@
+package uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.locations.availability
+
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.entity.AvailabilityStatus
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.Location
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.LocationUsage
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.response.AvailableLocation
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.response.AvailableLocationsResponse
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.model.slot
+import uk.gov.justice.digital.hmpps.hmppsbookavideolinkapi.service.locations.LocationsService
+import java.time.LocalTime
+
+abstract class LocationAvailabilityService<REQUEST>(private val locationsService: LocationsService) {
+
+  abstract fun findAvailable(request: REQUEST): AvailableLocationsResponse
+
+  protected fun getVideoLinkLocationsAt(prisonCode: String) = locationsService.getVideoLinkLocationsAtPrison(prisonCode = prisonCode, enabledOnly = true)
+
+  protected class AvailableLocationsBuilder private constructor() {
+    companion object {
+      fun availableLocations(init: AvailableLocationsBuilder.() -> Unit) = AvailableLocationsBuilder().also { it.init() }
+    }
+
+    private val dedicatedProbationTeamLocations = mutableSetOf<AvailableLocation>()
+    private val anyProbationTeamLocations = mutableSetOf<AvailableLocation>()
+    private val dedicatedCourtLocations = mutableSetOf<AvailableLocation>()
+    private val anyCourtLocations = mutableSetOf<AvailableLocation>()
+    private val sharedLocations = mutableSetOf<AvailableLocation>()
+
+    fun add(availabilityStatus: AvailabilityStatus, location: Location, startTime: LocalTime, endTime: LocalTime) {
+      AvailableLocation(
+        name = location.description ?: location.key,
+        startTime = startTime,
+        endTime = endTime,
+        dpsLocationId = location.dpsLocationId,
+        dpsLocationKey = location.key,
+        usage = location.extraAttributes?.locationUsage?.let { LocationUsage.valueOf(it.name) } ?: LocationUsage.SHARED,
+        timeSlot = slot(startTime),
+      ).let { availableLocation ->
+        when (availabilityStatus) {
+          AvailabilityStatus.PROBATION_TEAM -> dedicatedProbationTeamLocations.add(availableLocation)
+          AvailabilityStatus.PROBATION_ANY -> anyProbationTeamLocations.add(availableLocation)
```

Comment:
PROBATION_ROOM rather than PROBATION_TEAM?   (to mimic the COURT_ROOM below)?  Tiny, nit-picker only.